### DISCUSSION TITLE
Assumptions & questions on data sources/accuracy

### DISCUSSION THREAD

**Discussion by theguj** [theOehrly/Fast-F1#491] [Q&A]
Hi all, I'm newish to F1 (DTS generation) and very new to FastF1 and GitHub (first post), so please bear with me and beware my assumptions! I've been playing around with FastF1 for the last few weeks and have some thoughts/questions.

First off, thank you, @theOehrly!

The data F1 provides is great but there are also clearly gaps and inaccuracies. I would love feedback on my assumptions regarding Car and Position Data.

## Car Data _(get_car_data)_
**RPM**
- TLDR: Assume accurate enough for most analyses but not as accurate as the number of significant figures
- Unsure how it is measured (e.g. more direct or by engine sound)
- I had thought this might be a better way to estimate speed (since it's usually 4+ sig figures) but changes (within the same gear) are not as smooth as 'Speed' and I suspect it can only be trusted to 100s of RPMs

**Speed**
- TLDR: Assume relatively accurate (barring errors/edge cases, usually accurate to +/- 0.5 kmph)
- Speed values look reasonable when compared over time and imply reasonable accelerations (after accounting for precision)
- Does anyone know how Speed is measured?
  - I **assume** speed is the wheel speed calculated by measuring the revolutions downstream of the gearbox and then multiplying by "known" circumferences of the wheels? And not, based on GPS or Radar (like the speed trap)
  - I **assume** calculations account for dry vs. wet tyre sizes but not for wear or distortion from different pressures and fuel loads?
  - If correct, there will be edge cases such as wheel spin will overestimate a car's speed and other small errors
- For my analyses, I have adjusted speeds up/down (<0.5kmph) to give smooth linear acceleration between points but generally trusted

**nGear**
- TLDR: Assume mostly accurate (i.e. within gear) but lagging and inaccurate for the record prior to a shift
- Gear shifts take time (maybe order of 10s milliseconds but I cannot find a source, only that FIA says must be <200/300ms depending on up/down)
  - Does anyone know more accurately how quick a gear change is?
- Comparing throttles, gears and speeds, the data looks suspicious around gear changes (especially the row before a switch), so I suspect data may be lagging and when there is a shift it shows the gear prior to a completed shift
- For my analyses, I am replacing the row before a change with a shift indicator (e.g. '6>7')

**Throttle, Brake and DRS**
- Throttle: Assume accurate to significant figures (i.e. +/- 0.5%)
- Brake: Assume accurate, albeit without given any sense of how hard the brake is pressed
- DRS: Assume accurate and as documented in prior investigations (https://docs.fastf1.dev/_modules/fastf1/api.html)

**SessionTime**
- TLDR: Assume relatively accurate (i.e. +/- <1ms) and meaningful to synch with lap data
- Intervals seem to start at 0.16 but can be higher, and are multiples of 0.04 (e.g. 0.16, 0.20, 0.24...) but ~10% is +/- 0.001ms
  - Potentially data is logged on some sort of counter that works in ~0.04s cycles but not exactly?
- For my analyses, I am using its accurate for splitting sectors, calculating accelerations and interpolation
  - Note, I’m using SUVAT equations for interpolations, i.e. assuming linear acceleration for short time periods

## Position Data _(get_pos_data)_
**X, Y and Z**
- TLDR: Assume relatively accurate (i.e. +/- 0.5 or +/- 50mm)
- Anyone know how the position data is determined, I assume something using antennas around the track?
- I **assume** the X and Y values are pretty accurate, since even when looking at different laps and different drivers, the points generally fall very close to a racing line +/- 50mm
- No idea which part of the car it measures (hopefully it's close to or the same as the transponder used for determining lap starts/ends)
- Would be great to know angle of rotation but without it, I am estimating based on previous and subsequent records
- Position data would be much more useful with an accurate set of coordinates for Start/Finish and Sector lines but in lieu of these I’m estimating (see approach at the bottom)

**SessionTime**
- TLDR: Assume quite inaccurate (i.e. often +/- 100ms)
- By looking at the distance between coordinates and comparing to the time passed, we can calculate velocity. Overall, the values make sense but there are many outliers. Initially, I assumed the positions were the issue but as mentioned above, they look reasonable
- I **assume** there are errors and that there are due to some sort of inconsistent delays
- Unfortunately, this means that not only is it quite risky to merge Car and Position data based on SessionTime but that even individual records near the start/finish line are often going to be allocated to the wrong lap
  - I am working on the assumption that most of the time SessionTime is fairly accurate but that (somewhat randomly) it could be significantly off (see approach at the bottom)

## Approach for merging car and position data
- Most of the time, merging on SessionTime should be fine but I am hoping to do relatively detailed corner analyses. Given the likely issues with Position:SessionTime, I am opting to merge based on ‘EstDistance’
- Getting Car Data ‘EstDistance’
  - Assuming Speed is in the direction of travel, distance covered between records is the average speed multiplied by the time between records
  - Some approaches use a single speed but there is almost continuous speed changes, so I am assuming constant acceleration (between records) and taking the average speed (effectively S = (U + V) * T / 2
  - I then take the cumulative distance since the lap start (extrapolating the start using SUVAT)
  - This approach is still imperfect but it’s likely pretty good, since total lap distances are usually within a few metres of each other (plus in reality drivers will take different lines so distance **are** truly different)
  - However, I want to compare drivers at particular corners, so I am rescaling all distances
    - First, I am applying the SUVAT equations to interpolate to find distances as sector lines
    - Then, I’m taking a median across all laps/drivers for a circuit to get a distance between sectors and scaling the distances calculated to that median
- Getting Position Data ‘EstDistance’ **[Not done, this is my plan]**
  - Distances between records can be found using Pythagoras’ theorem but the challenge is knowing the distance from the start
  - If SessionTime could be trusted, we could apply SUVAT to calculate the lap start but different laps/drivers are giving different starts because of SessionTime errors
  - I am assuming that while there are many errors, after removing outliers, the average SessionTime should be fairly accurate and most records are in the correct lap
  - Based on determining a start set of X and Y coordinates, I can draw a line and determine which records are in the wrong lap
  - I can calculate EstDistance for the Position Data
- I will the merge Car Data and Position Data based on EstDistance not SessionTime
 - Although I have not completed the Distance cleaning, I merged based on a more crude distance covered and it worked pretty well

Thanks and congrats if you read all/much of that! Let me know what you’ve found, your thoughts on my assumptions and any ideas on approaches.

**Comment by theOehrly** [comment]
Hi, I'm not sure how to respond here without writing half a book.  In general, if I knew the answers to most of your questions, they'd be documented somewhere. In fact, no one here really knows the details about what you are asking. I'll pick out a few specifics that I do want to try and answer.

- RPM, speed, gear, throttle are most likely pulled directly from the ECU (so not externally measured). How the speed is measured on the car, no idea. Probably not (just) wheel speed, else you'd likely see lock ups in the speed trace.
- no idea whether the speed measurement accounts for wet/dry tire diameter difference
- no idea how accurate throttle is
- SessionTime is not guranteed to be accurate to +-1ms in the raw data. What you are seeing is the correction that FastF1 applies using the lap and sector times. This can generally only be done to correct SessionTimes for each driver but not across drivers. In race-like sessions, the gap to leader data is additionally used, to synchronize timing data across drivers. But there is still no absolute reference point, so everything may be shifted by a constant value. Synchronization adjusts everything to the value with the lowest delay, but it is not possible to say how large that delay still is.
- In the telemetry data you can observe the actual inaccuracies of the SessionTime because here, it isn't possible to apply corrections (at least I haven't yet found a reliable way)
- position data has an unknown source. It looks like position is given as a normalized position to a predefined line. Different lines and track excursions cannot be observed. That means you can't analyse line choices through a corner.

In general, you should always remember that was not originally intended for technical analysis. It was (is) intended for live visualization where delays of a few 100ms do not matter as they are barely noticeable by eye. The raw session time is something like packet timestamps with an unknown delay that varies for each packet. These can be corrected partially (in the case of timing data).

I can't really comment on your idea for merging and 'EstDistance'. Given the low sample rate of the data and the existing inaccuracies, things that should work in theory often do not work well in practice. Therefore, trial and error is often the only way forward.

**Comment by theguj** [comment]
Yeah, it was a bit of an essay! Thanks for the response!

As you say, the intent of the data is not for detailed analyses, so I probably shouldn't be surprised there are so many unknowns. Given how widespread the use of the data and your tools are, it's a shame there's not more official/inside knowledge around. I would guess that media and the teams do get more explanations (and potentially higher accuracy data). I would also assume that F1 are aware of a lot of the X, YouTube and Medium analyses using this data but it's not in their interests to support further.

Picking up on your bullets:
- You might be right about Speed being more complex/different than calculated wheel speed. Although, re lockups, I would guess lockups are usually front wheels, not the driven wheels at the back, while I would assume measurements would be from what's driving the rear wheels. Dunno though. Probably tricky but potentially I can infer some things from looking at edge cases (lockups, wheel-spins, doughnuts etc.)
- Accounting for wet/dry was just a guess. I think commentators once said Gasly was penalised for pit land speeding because he'd forgotten to set a button to indicate he was on wet tyres, so his limiter underestimated his true speed. That made me suspect that F1 is gathering the same calculations used by the limiter
- Thanks for elaborating on the SessionTime corrections. I'll read through your code a bit more. I think you're saying that the Car Data relative times (for a given driver/session) should be fairly accurate, which is good enough for my purposes
- That makes sense and would explain why it looks "too good to be true". I'm only trying to get distances relative to the track (start, sectors, DRS, corners) so it should be good for my purposes

I'll post again with how I get on. There's a big risk that I make everything look good with my various "corrections" but actually degrade the accuracy. Fingers crossed!
